Sources/StreamChat/Controllers/EventsController/EventsController.swift

```diff
@@ -0,0 +1,154 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+public extension _ChatClient {
+    func createEventsController() -> EventsController {
+        .init(notificationCenter: eventNotificationCenter)
+    }
+}
+
+public class EventsController: Controller {
+    public var callbackQueue: DispatchQueue = .main
+    private let notificationCenter: EventNotificationCenter
+    private var observers: [EventObserver] = []
+    
+    init(notificationCenter: EventNotificationCenter) {
+        self.notificationCenter = notificationCenter
+    }
+    
+    // MARK: - System events
+    
+    public func subscribe<T: Event>(
+        on: T.Type = T.self,
+        filter: @escaping (T) -> Bool = { _ in true },
+        handler: @escaping (T) -> Void
+    ) {
+        observers += [
+            .init(
+                notificationCenter: notificationCenter,
+                transform: { $0 as? T },
+                callback: { [unowned self] event in
+                    guard filter(event) else { return }
+                    self.callback { handler(event) }
+                }
+            )
+        ]
+    }
+    
+    // MARK: - Custom events
+    
+    public func subscribe<T: CustomEvent>(
+        on: T.Type = T.self,
+        filter: @escaping (T) -> Bool = { _ in true },
+        handler: @escaping (T) -> Void
+    ) {
+        subscribe(on: AnyCustomEvent.self) {
+            guard let event = $0.event(ofType: T.self), filter(event) else { return }
+            handler(event)
+        }
+    }
+    
+    public func send<T: CustomEvent>(
+        event: T.Type = T.self,
+        completion: @escaping (Error?) -> Void
+    ) {
+        // TODO:
+    }
```

Comment:
@tbarbugli @VojtaStavik what is the endpoint for sending custom events? Is it possible to manually initiate/send a system event?

Reply:
decided to handle custom events in separate PR